pkg/envoy/restart/handler.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package restart
+
+import (
+	"istio.io/istio/pkg/envoy"
+)
+
+// Handler of events from an Envoy manager Manager
+type Handler interface {
+	// OnStarting is called when the Manager is starting the run loop.
+	OnStarting()
+
+	// OnShuttingDown is called when the Manager has begun a graceful shutdown.
+	OnShuttingDown()
+
+	// OnStopped is called when the Manager has exited the run loop.
+	OnStopped(err error)
+
+	// OnEnvoyStarting is called when the Manager has started a new Envoy server.
+	OnEnvoyStarting(e envoy.Instance)
+
+	// OnEnvoyStopped is called when an Envoy server has terminated. If the returned
+	// error is context.Canceled, Envoy was internally aborted by the Manager.
+	OnEnvoyStopped(e envoy.Instance, err error)
+}
+
+// CompositeHandler is a Handler that can augmented with additional handler functions.
+type CompositeHandler interface {
+	Handler
+	WithOnStarting(fns ...func()) CompositeHandler
+	WithOnShuttingDown(fns ...func()) CompositeHandler
+	WithOnStopped(fns ...func(error)) CompositeHandler
+	WithOnEnvoyStarting(fns ...func(envoy.Instance)) CompositeHandler
+	WithOnEnvoyStopped(fns ...func(envoy.Instance, error)) CompositeHandler
+}
+
+// NewCompositeHandler creates a new CompositeHandler.
+func NewCompositeHandler() CompositeHandler {
+	return &handlerImpl{}
+}
+
+type handlerImpl struct {
+	onStarting      []func()
+	onShuttingDown  []func()
+	onStopped       []func(error)
+	onEnvoyStarting []func(envoy.Instance)
+	onEnvoyStopped  []func(envoy.Instance, error)
+}
+
+func (h *handlerImpl) OnStarting() {
+	for _, f := range h.onStarting {
+		f()
+	}
+}
+
+func (h *handlerImpl) OnShuttingDown() {
+	for _, f := range h.onShuttingDown {
+		f()
```

Comment:
do we need this abstraction here? I tend to find this impossible to figure out what is actually being called in set ups like this. Can we just do 
```
func OnStarting() {
   startfn1()
   startfn2()
}
```
and keep it explicit? 

I don't see where this is actually being called either (probably just github being annoying)

Reply:
+1. Please don't invert the control flow, it makes very much unreadable.

Reply:
This is a composite handler, so each event has `n` handler functions that need be notified. All we're doing here is iterating through the appropriate event handlers.

If the concern is simply the idea of a composite handler, I'm fine removing it for now.

sg?

Reply:
I just want to be able to read the follow the code by following the function calls. Every time there's some function handler that is assigned separately, it makes it harder to follow.

Reply:
So the requirement here is to allow the caller to provide independent handler for lifecycle events, so I'm not that we can eliminate some abstraction here.

Reply:
I am not sure I understand. We aren't building a generic library here that will be used by many users with arbitrary call goals -- aren't we are the only users of this? Can we just replace this with 

```go
func OnStarting() {
			m.eventCh <- Event{
				Type: Starting,
			}
}
```

(and if thats not actually the code that is called from OnStarting then that is another sign that its a bit hard to follow)

Reply:
IMO, the lack of extensibility here is how we ended up with the agent kludge we have today. There's no clear separation of what's core to running Envoy and what are extensions. As a result, pilot-agent expands to be the tightly coupled monolith it is today. This PR moves towards a more composable and modular pilot-agent.

Reply:
The implementation details of how we do this should definitely be discussed though!

Reply:
If we want to make it extensible then it seems we could do stuff like

```go
func OnStarting() {
   startEnvoy()
   startMonitoring()
   startMaintenance()
}
```

if I am understanding the goals here correctly?

I think we can make it clean without being this abstract. Its really hard to tell what is going on here. I am still not 100% sure what exactly is called from `OnStarting` even after pulling it into my IDE, and similar patterns across the codebase have been a challenge to debug in the past.

Reply:
@kyessenov What were the upstreaming requirements you had in the preceding PR?